server/server.js

```diff
@@ -0,0 +1,264 @@
+// @flow
+import type {IncomingMessage, ServerResponse} from 'http'
+import type {NextConfig} from './config'
+
+import {STATUS_CODES} from 'http'
+import {parse as parseUrl} from 'url'
+import {parse as parseQuerystring} from 'querystring'
+import {resolve, join, sep} from 'path'
+import send from 'send'
+import {PHASE_PRODUCTION_SERVER, PHASE_DEVELOPMENT_SERVER} from '../lib/constants'
+import getConfig from './config'
+import * as asset from '../lib/asset' // can be deprecated
+
+type ServerConstructor = {
+  dir?: string,
+  dev?: boolean,
+  quiet?: boolean,
+  conf?: NextConfig
+}
+
+type UrlParseResult = {
+  protocol?: string;
+  slashes?: boolean;
+  auth?: string;
+  host?: string;
+  port?: string;
+  hostname?: string;
+  hash?: string;
+  search?: string;
+  query?: any; // null | string | Object
+  pathname?: string;
+  path?: string;
+  href: string;
+}
+
+type ParsedUrl = {
+  protocol?: string;
+  slashes?: boolean;
+  auth?: string;
+  host?: string;
+  port?: string;
+  hostname?: string;
+  hash?: string;
+  search?: string;
+  query: any; // null | string | Object
+  pathname: string;
+  path: string;
+  href: string;
+}
+
+type Envelope = {
+  statusCode: number,
+  content?: string,
+  headers?: {[key: string]: string},
+  filePath?: string
+}
+
+async function serveFilePath (req: IncomingMessage, res: ServerResponse, filePath: string) {
+  return new Promise((resolve, reject) => {
+    send(req, filePath)
+      .on('directory', () => {
+        // We don't allow directories to be read.
+        resolve(false)
+      })
+      .on('error', () => {
+        resolve(false)
+      })
+      .pipe(res)
+      .on('finish', () => {
+        resolve(true)
+      })
+  })
+}
+
+async function sendEnvelope (req: IncomingMessage, res: ServerResponse, {statusCode, headers, content, filePath}: Envelope): Promise<void> {
+  res.statusCode = statusCode
+
+  if (headers) {
+    for (const header of Object.keys(headers)) {
+      res.setHeader(header, headers[header])
+    }
+  }
+
+  // If a filePath is provided we serve the file contents instead of the `content` property
+  if (filePath) {
+    const servedFile = await serveFilePath(req, res, filePath)
+    if (!servedFile) {
+      // If the file couldn't be served we return a 404 status
+      res.statusCode = 404
```

Comment:
right now we're saying "if any error happens when reading the file (from the `serveFilePath` code), it's a 404. That's not right. It could be a bunch of issues

Reply:
For instance, someone could have permission issues with the file and will spend a lot of time debugging trying to understand this error